client/material/text-field.tsx

```diff
@@ -0,0 +1,416 @@
+  $dense?: boolean
+  $disabled?: boolean
+  $floatingLabel?: boolean
+  $focused?: boolean
+  $maxRows?: number
+  $multiline?: boolean
+}>`
+  display: flex;
+  flex-direction: column;
+  align-items: flex-start;
+  position: relative;
+  width: 100%;
+  min-height: ${props => (props.$dense ? '40px' : '56px')};
+  font-size: 16px;
+  line-height: 20px;
+  background-color: rgba(255, 255, 255, 0.1);
+  border-radius: 4px 4px 0 0;
+  contain: layout paint style;
+  user-select: text;
+
+  ${props => {
+    const spacing = props.$floatingLabel ? 34 : 28 /* textfield padding + input margin */
+    const height = props.$dense ? 40 : 56
+
+    return `max-height: ${props.$maxRows ? props.$maxRows * 20 + spacing : height}px;`
+  }}
+
+  ${props => {
+    if (!props.$multiline) return ''
+
+    if (props.$floatingLabel) {
+      return props.$dense ? 'padding: 17px 0 2px 12px' : 'padding: 25px 0 2px 12px;'
+    } else {
+      return props.$dense ? 'padding: 11px 0 2px 12px' : 'padding: 19px 0 2px 12px;'
+    }
+  }}
+
+  ${props =>
+    !props.$disabled
+      ? `
+        &:hover {
+          background-color: rgba(255, 255, 255, 0.12);
+        }
+      `
+      : ''}
+
+  ${props =>
+    props.$focused
+      ? `
+        background-color: rgba(255, 255, 255, 0.16) !important;
+      `
+      : ''}
+
+  & input[type='text'],
+  & input[type='password'],
+  & input[type='datetime'],
+  & input[type='datetime-local'],
+  & input[type='date'],
+  & input[type='month'],
+  & input[type='time'],
+  & input[type='week'],
+  & input[type='number'],
+  & input[type='email'],
+  & input[type='url'],
+  & input[type='search'],
+  & input[type='tel'],
+  & input[type='color'],
+  & textarea {
+    -moz-appearance: none;
+    -webkit-appearance: none;
+  }
+`
+
+const iconStyle = css<{ $dense?: boolean }>`
+  position: absolute;
+  top: 4px;
+  width: ${props => (props.$dense ? '32px' : '48px')};
+  height: ${props => (props.$dense ? '32px' : '48px')};
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  color: ${colorTextSecondary};
+`
+
+const LeadingIcon = styled.span<{ $dense?: boolean; $index: number }>`
+  ${iconStyle}
+  left: ${props => {
+    const iconWidth = props.$dense ? 32 : 48
+    const leftOffset = props.$index * iconWidth + (props.$index + 1) * 4
+
+    return `${leftOffset}px`
+  }};
+`
+
+const TrailingIcon = styled.span<{ $dense?: boolean; $index: number; $multiline?: boolean }>`
+  ${iconStyle}
+  right: ${props => {
+    const iconWidth = props.$dense ? 32 : 48
+    const multilinePadding = props.$multiline ? 12 : 0
+    const rightOffset = props.$index * iconWidth + (props.$index + 1) * 4 + multilinePadding
+
+    return `${rightOffset}px`
+  }};
+`
+
+const ClearTooltip = styled(Tooltip)`
+  height: 32px;
+`
+
+const ClearButton = styled(IconButton)`
+  width: 32px;
+  min-height: 32px;
+`
+
+export type TextSelectionDirection = 'forward' | 'backward' | 'none'
+
+export interface TextFieldHandle {
```

Comment:
This should most definitely not exist, just pass back a ref to the input. Otherwise we have to forward every single method/value/whatever that we think is vaguely useful. The only thing in here that isn't a direct forward is `clear`, and `clear` makes no sense to have here (this component is controlled, if you want to clear it, you should be setting the new value via props)